speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/FileDataStorage.cpp

```diff
@@ -98,64 +121,285 @@ void FileDataStorage::Shutdown()
+    }
+
     if (p_decisions.ContainsGear())
     {
-        WRITE_STRING(m_outputStream, "GearDecision");
-        WRITE_VAR(m_outputStream, p_decisions.GetGear());
+        AddIntToArray(m_gearDecision, p_decisions.GetGear(), m_compressionStep);
+    }
+    else
+    {
+        AddIntToArray(m_gearDecision, -1, m_compressionStep);
     }
+
     if (p_decisions.ContainsLights())
     {
+        AddIntToArray(m_lightDecision, p_decisions.GetLights(), m_compressionStep);
+    }
+    else
+    {
+        AddIntToArray(m_lightDecision, -1, m_compressionStep);
+    }
+}
+
+void FileDataStorage::WriteCarData()
+{
+    WRITE_VAR(m_outputStream, GetAverage(m_totalPosX));       // x-position
+    WRITE_VAR(m_outputStream, GetAverage(m_totalPosY));       // y-position
+    WRITE_VAR(m_outputStream, GetAverage(m_totalPosZ));       // z-position
+    WRITE_VAR(m_outputStream, GetAverage(m_totalPosAx));      // x-direction
+    WRITE_VAR(m_outputStream, GetAverage(m_totalPosAy));      // y-direction
+    WRITE_VAR(m_outputStream, GetAverage(m_totalPosAz));      // z-direction
+    WRITE_VAR(m_outputStream, GetAverage(m_totalMovVelX));    // speed
+    WRITE_VAR(m_outputStream, GetAverage(m_totalMovAccX));    // acceleration
+    WRITE_VAR(m_outputStream, GetLeastCommon(m_gearValues));  // gear
+}
+
+void FileDataStorage::WriteHumanData()
+{
+    WRITE_VAR(m_outputStream, GetMedian(m_steerValues));   // steer
+    WRITE_VAR(m_outputStream, GetMedian(m_brakeValues));   // brake
+    WRITE_VAR(m_outputStream, GetMedian(m_accelValues));   // gas
+    WRITE_VAR(m_outputStream, GetMedian(m_clutchValues));  // clutch
+}
+
+void FileDataStorage::WriteInterventionData()
+{
+    bool decisionMade = false;
+    float decision = GetMedian(m_steerDecision);
+    if (decision != -1)
+    {
+        decisionMade = true;
+        WRITE_STRING(m_outputStream, "Decisions");
+        WRITE_STRING(m_outputStream, "SteerDecision");
+        WRITE_VAR(m_outputStream, decision);
+    }
+    decision = GetMedian(m_brakeDecision);
+    if (decision != -1)
+    {
+        if (!decisionMade)
+        {
+            decisionMade = true;
+            WRITE_STRING(m_outputStream, "Decisions");
+        }
+        WRITE_STRING(m_outputStream, "BrakeDecision");
+        WRITE_VAR(m_outputStream, decision);
+    }
+    decision = GetMedian(m_accelDecision);
+    if (decision != -1)
+    {
+        if (!decisionMade)
+        {
+            decisionMade = true;
+            WRITE_STRING(m_outputStream, "Decisions");
+        }
+        WRITE_STRING(m_outputStream, "AccelDecision");
+        WRITE_VAR(m_outputStream, decision);
+    }
+    int intDecision = GetLeastCommon(m_gearDecision);
+    if (intDecision != -1)
+    {
+        if (!decisionMade)
+        {
+            decisionMade = true;
+            WRITE_STRING(m_outputStream, "Decisions");
+        }
+        WRITE_STRING(m_outputStream, "GearDecision");
+        WRITE_VAR(m_outputStream, intDecision);
+    }
```

Comment:
You could look into extracting this to a define/function since it has quite some repeated code